### PR TITLE
feat(derive):  implement the FromEvent trait and derive macro

### PR DESCRIPTION
**Pull request by joshuachp**
Implement a trait to convert an `AstarteDataEvent` into a rust struct. Since the implementation is repetitive and has to pass some standard checks, a derive macro is also implemented for named structs. It will parse the fields of the struct and extract the appropriate endpoints values from the event.

### CODE REVIEW

**Comment by joshuachp** [comment]
This pull request depends on #236

File: astarte-device-sdk-derive/src/case.rs [hunk 1]
```diff
@@ -72,10 +76,12 @@ impl RenameRule {
     }
 
     /// Apply a renaming rule to a struct field, returning the version expected in the source.
-    pub fn apply_to_field(&self, field: &str) -> String {
+    pub fn apply_to_field<'a>(&self, field: &'a str) -> Cow<'a, str> {
         match *self {
-            RenameRule::None | RenameRule::LowerCase | RenameRule::SnakeCase => field.to_owned(),
-            RenameRule::UpperCase => field.to_ascii_uppercase(),
+            RenameRule::None | RenameRule::LowerCase | RenameRule::SnakeCase => {
```
Comment (rgallor) [inline astarte-device-sdk-derive/src/case.rs:81]: LowerCase and SnakeCase assume that the field is following Rust's field naming convention

**Review by rgallor — CHANGES_REQUESTED** [review]

**Review by lucaato — APPROVED** [review]

**Review by harlem88 — CHANGES_REQUESTED** [review]

**Review by harlem88 — COMMENTED** [review]
Add an entry to the changelog about this new feature.

File: src/event.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+///
+///         let name = object
+///             .remove("name")
+///             .ok_or(FromEventError::MissingField {
+///                 interface: "com.example.Sensor",
+///                 base_path: "sensor",
+///                 path: "name",
+///             })?
+///             .try_into()?;
+///         let value = object
```
[the remark this answers is not in the document]
Comment (joshuachp) [reply src/event.rs:112]: Done

File: astarte-device-sdk-derive/src/lib.rs [hunk 3]
```diff
@@ -266,9 +448,95 @@ fn parse_astarte_aggregate_attribute(attr: &Attribute) -> Option<syn::Result<Agg
         syn::Meta::Path(_) => None,
```
[the remark this answers is not in the document]
Comment (joshuachp) [reply astarte-device-sdk-derive/src/lib.rs:448]: Done

**Comment by joshuachp** [comment]
> Add an entry to the changelog about this new feature.

Added

File: astarte-device-sdk-derive/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,9 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Feat
+- Add macro to implement the `FromEvent` trait on a generic struct.
```
Comment (harlem88) [inline astarte-device-sdk-derive/CHANGELOG.md:9]: We follow [keepachangelog](https://keepachangelog.com/en/1.0.0/#how), use `Added` for introducing a new features.
```suggestion
### Added
- Macro to implement the `FromEvent` trait on a generic struct.
```

Comment (joshuachp) [reply astarte-device-sdk-derive/CHANGELOG.md:9]: My bad, fixed

**Review by rgallor — APPROVED** [review]

**Review by harlem88 — APPROVED** [review]

**Pull request merged**